src/libraries/System.Linq/src/System/Linq/Max.cs

```diff
@@ -8,157 +8,70 @@ namespace System.Linq
-                {
-                    value = span[i];
-                }
-            }
-
-            return value;
-        }
-
-        public static int? Max(this IEnumerable<int?> source)
-        {
-            if (source == null)
-            {
-                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
-            }
+                    while (index + Vector<T>.Count <= span.Length);
 
-            int? value = null;
-            using (IEnumerator<int?> e = source.GetEnumerator())
-            {
-                do
-                {
-                    if (!e.MoveNext())
+                    value = maxes[0];
+                    for (int i = 1; i < Vector<T>.Count; i++)
                     {
-                        return value;
-                    }
-
-                    value = e.Current;
-                }
-                while (!value.HasValue);
-
-                int valueVal = value.GetValueOrDefault();
-                if (valueVal >= 0)
-                {
-                    // We can fast-path this case where we know HasValue will
-                    // never affect the outcome, without constantly checking
-                    // if we're in such a state. Similar fast-paths could
-                    // be done for other cases, but as all-positive
-                    // or mostly-positive integer values are quite common in real-world
-                    // uses, it's only been done in this direction for int? and long?.
-                    while (e.MoveNext())
-                    {
-                        int? cur = e.Current;
-                        int x = cur.GetValueOrDefault();
-                        if (x > valueVal)
+                        if (maxes[i] > value)
                         {
-                            valueVal = x;
-                            value = cur;
+                            value = maxes[i];
                         }
                     }
                 }
                 else
                 {
-                    while (e.MoveNext())
-                    {
-                        int? cur = e.Current;
-                        int x = cur.GetValueOrDefault();
+                    value = span[0];
+                    index = 1;
+                }
 
-                        // Do not replace & with &&. The branch prediction cost outweighs the extra operation
-                        // unless nulls either never happen or always happen.
-                        if (cur.HasValue & x > valueVal)
-                        {
-                            valueVal = x;
-                            value = cur;
-                        }
+                // Iterate through the remaining elements, comparing against the max.
+                for (int i = index; (uint)i < (uint)span.Length; i++)
+                {
+                    if (span[i] > value)
+                    {
+                        value = span[i];
                     }
```

Comment:
I've done various checks for code I've looked at locally, but it would be nice to validate there aren't any codegen changes for the scalar fallback path here.

Would help indicate if there is a need to tweak the inliner at all for these scenarios.

Reply:
CC. @EgorBo. Might be interesting to give generic math operators an inline multiplier, particularly if small or for the primitive types (which are normally single inline instructions)

Reply:
> it would be nice to validate there aren't any codegen changes for the scalar fallback path here.

This is Max on `IEnumerable<int?>`... it looks ok to me:
https://www.diffchecker.com/MU3Kg6ow